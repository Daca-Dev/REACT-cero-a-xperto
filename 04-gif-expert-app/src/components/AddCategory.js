import React, { useState } from 'react';
import PropTypes from 'prop-types'


const AddCategory = ({ setCategories }) => {


    // se deben de usar useStates para poder cambiar y actualizar los valores de variables en el rende de react
    const [inputValue, setInputValue] = useState('');

    const handleInput = (event) => {

        setInputValue( event.target.value );

    }

    const handleSubmit = (event) => {
        /** manejar el evento de submit (presionar enter) en el formulario */
        // evita que el formulario recargue la página, esto ya no se debería aher así
        event.preventDefault();

        if ( inputValue.trim().length > 2 ) {
            // uso de la función pasada en los proptypes por medio de un callback
            setCategories( c => [ inputValue, ...c,] );
            setInputValue('');
        }
        
    }

    return (
        <form onSubmit={ handleSubmit }>
            <input
                type="text"
                value={ inputValue }
                onChange={ handleInput }
                placeholder='Ingrese un texto de búsqueda...'
            />
        </form>
    )
}

// configuracion de los proptypes
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}


export default AddCategory
