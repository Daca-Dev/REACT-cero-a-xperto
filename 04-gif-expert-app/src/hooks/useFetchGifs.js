import { useState, useEffect } from "react";
import { getGifs } from "../helpers/getGifs";


export const useFetchGifs = ( category ) => {
    
    const [ state, setState ] = useState({
        data: [], loading: true
    })

    // use efect es usado para controlar cunado se deben de ejcutar piezas de cÃ³digo cuando
    // se renderizan los componentes cada vez que hay un cambio
    useEffect( () => {

        getGifs( category )
            .then( imgs => {
                setTimeout(() => {
                    setState({
                        data: imgs,
                        loading: false
                    });
                }, 2000);
            })
            .catch( err => console.log( err ));
    }, [ ]);

    return state
}



