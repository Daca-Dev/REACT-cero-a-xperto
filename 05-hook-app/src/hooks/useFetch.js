import { useState, useEffect, useRef } from 'react';


export const useFetch = ( url ) => {


    const isMount = useRef( true );
    const [ state, setState ] = useState({
        loading: true,
        error: null,
        data: null
    })

    useEffect( () => {
        console.log('Montado');
        return () => {
            isMount.current = false;
            console.log('Desmontado');
        }
    }, [ ] );

    useEffect( () => {

        setState({ loading: true, error: null, data: null })

        fetch( url )
            .then( resp => resp.json() )
            .then( data => {

                setTimeout(() => {
                    if ( isMount.current ) {
                        setState({
                            loading: false,
                            error: null,
                            data
                        })
                    } else {
                        console.log('SetState no llamado');
                    }
                }, 2000);
            })

    }, [ url ])

    return state;
}