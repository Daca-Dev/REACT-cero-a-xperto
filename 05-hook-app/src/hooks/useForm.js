import { useState } from "react";

export const useForm = ( initialForm ) => {

    const [ form, setForm ] = useState( initialForm );

    const resetForm = () => {
        setForm( initialForm );
    }

    const handleOnChange = ( { target } ) => {

        setForm({
            ...form,
            [ target.name ]: target.value
        })
    }

    return [ form, handleOnChange, resetForm ];

}