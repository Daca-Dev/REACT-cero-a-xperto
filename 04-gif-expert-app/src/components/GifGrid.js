import React from 'react'
import PropTypes from 'prop-types'
import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';


export const GifGrid = ({ category }) => {

    
    const { data: images, loading } = useFetchGifs( category );

    return (
        <div className="animate__animated animate__flipInX">
            <h3>{ category }</h3>
            <div className="cards-grid">
                { loading && <p>...Loading</p>}
                {
                    images.map( img => 
                        <GifGridItem 
                            key={ img.id }
                            { ...img }
                        />
                    )
                }
            </div>
        </div>
    )
}


GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}