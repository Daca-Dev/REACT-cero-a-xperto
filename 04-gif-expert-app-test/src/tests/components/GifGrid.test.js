import React from 'react';
import '@testing-library/jest-dom';
import {shallow} from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');



describe('Pruebas del componente <GifGrid />', () => {
    
    const category = 'halo';

    test('El componente debe de renderizar correctamente', () => {

        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow( <GifGrid category={ category } />);
        expect( wrapper ).toMatchSnapshot();
    })

    test('Debe de mostrar items cuando se cargan imagenes useFetchGifs', () => {
        // simular la data que recibe

        const gif = [{
            id:'ABC',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: 'Cualqueri cosa'
        }]

        useFetchGifs.mockReturnValue({
            data: gif,
            loading: false
        });
        
        const wrapper = shallow( <GifGrid category={ category } />);
        
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBe( false );
        expect( wrapper.find( 'GifGridItem' ).length ).toBe( gif.length );

    })
    

})
