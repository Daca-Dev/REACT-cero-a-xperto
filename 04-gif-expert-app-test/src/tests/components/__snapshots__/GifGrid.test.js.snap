// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente <GifGrid /> Debe de mostrar items cuando se cargan imagenes useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
     
    halo
     
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualqueri cosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
  </div>
</Fragment>
`;

exports[`Pruebas del componente <GifGrid /> El componente debe de renderizar correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
     
    halo
     
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
